### PR TITLE
[WIP] FIX: setting requires_grad on adapter layers

### LINKED ISSUE #899: Modules_to_save can't be used with DDP and gradient checkpointing
### System Info

To train a model with gradient checkpointing and DDP, the model should have no unused parameters and the HF trainer's ddp_find_unused_parameters argument should be set to False.

However, peft's modules_to_save functionality creates unused parameters: the ModulesToSaveWrapper's original_module.

Gradient_checkpointing and DDP are really powerful in peft runs since large models can be made to fit on a single GPU, so it would be great to fix this functionality.

I attach a minimal example of a PEFT training script that fails with ddp.

### Who can help?

@pacman100 @younesbelkada @BenjaminBossan 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer, AutoModelForCausalLM, Trainer
from peft import (
    TaskType,
    LoraConfig,
    get_peft_model,
    prepare_model_for_kbit_training
)
from transformers import DataCollatorForSeq2Seq, TrainingArguments
from datasets import load_dataset

model = AutoModelForCausalLM.from_pretrained('EleutherAI/pythia-70m',device_map='auto',torch_dtype='auto')
tokenizer = AutoTokenizer.from_pretrained('EleutherAI/pythia-70m')
tokenizer.pad_token_id = 0

model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=True)

model = get_peft_model(model, LoraConfig(
    task_type=TaskType.CAUSAL_LM,
    inference_mode=False,
    r=8,
    lora_alpha=16,
    lora_dropout=0.1,
    bias="none",
    target_modules=['dense'],
    modules_to_save=['embed_out'],
)

dataset = load_dataset('tatsu-lab/alpaca', split="train[:100]")

train_data = dataset.map(lambda x: tokenizer(x['text'], return_tensors=None, padding='max_length', truncation=True, max_length=128))
train_data = train_data.map(lambda x: {'labels': x['input_ids']})

data_collator = DataCollatorForSeq2Seq(
    tokenizer=tokenizer,  pad_to_multiple_of=8, return_tensors="pt", padding=True
)

trainer = Trainer(
    model=model,
    train_dataset=train_data,
    args=TrainingArguments(output_dir='dummy', 	 
          per_device_train_batch_size=1,
          save_strategy='epoch',
          save_steps=1,
          ddp_find_unused_parameters=False),
    data_collator=data_collator)

trainer.train()
```

DDP launch command:
```bash
torchrun \
--nproc_per_node=1 \
--nnodes=1 \
--rdzv-backend=c10d \
--rdzv-endpoint=localhost:0 \
test.py
```

### Expected behavior

The training run should launch

### PR DESCRIPTION
**Pull request by BenjaminBossan**
This is an alternative to #900, resolves #899.

Thanks @passaglia for figuring out the underlying issue.

## Description

Currently, we don't handle setting `requires_grad` on adapter layers really well. The main issue is that it can be set to `True` on adapter parameters that are not being used, e.g. the `original_module` in `ModulesToSaveWrapper` or inactive adapters in LoRA.

Normally, this is not a big issue, except maybe if we want to correctly count the number of trainable parameters. However, when training with `DistributedDataParallel`, this results in errors, as PyTorch thinks that all parameters with `requires_grad=True` should participate in the loss computation, but those mentioned parameters don't. For that reason, training with DDP currently errors when using `modules_to_save` or multiple adapters.

## Implementation

This turned out to be more complicated than I initially thought. The logic for setting `requires_grad` is all over the place, it was hard to encapsulate the logic and I only succeeded partially. As is, this PR is more complex than the one it tries to supersede, #900, but it is also "more correct".

Tests were added to check whether `requires_grad` is set correctly. There are (so far) no tests for whether DDP indeed works, they could be added with multi-GPU. I did, however, test an early stage of this PR with DDP and setting `requires_grad` correctly will indeed fix the DDP error.

## DONE/TODO

- [x] ModulesToSaveWrapper
- [x] LoRA
- [ ] IA³
- [ ] AdaLora

Since some tuners are not implemented yet, tests are expected to fail. Check the new tests at the bottom of `test_custom.py`, those should pass. I have not measured if this change adds a significant overhead when switching adapters.

### CODE REVIEW

**Review by pacman100 — APPROVED** [review]
Thank you @BenjaminBossan for fixing this major bug when using DDP/Multiple Adapters with PEFT. LGTM! 🤗

**Review by younesbelkada — APPROVED** [review]
Thanks a mile @BenjaminBossan !
